Stratis.Bitcoin.Features.Consensus/ConsensusLoop.cs

```diff
@@ -62,78 +164,197 @@ public void Initialize()
+                            return;
+                    }
 
-                this.AcceptBlock(new ContextInformation(result, this.Validator.ConsensusParams));
+                    // TODO Need to revisit unhandled exceptions in a way that any process can signal an exception has been
+                    // thrown so that the node and all the disposables can stop gracefully.
+                    this.logger.LogCritical(new EventId(0), ex, "Consensus loop at Tip:{0} unhandled exception {1}", this.Tip, ex.ToString());
+                    NLog.LogManager.Flush();
+                    throw;
+                }
             }
-            catch (ConsensusErrorException ex)
+        }
+
+        /// <summary>
+        /// A method that will accept a new block to the node.
+        /// The block will be validated and the <see cref="CoinView"/> db will be updated.
+        /// If it's a new block that was mined or staked it will extend the chain and the new block will set <see cref="ConcurrentChain.Tip"/>.
+        /// </summary>
+        /// <param name="item">The block to validate.</param>
+        public void AcceptBlock(BlockItem item)
+        {
+            this.logger.LogTrace("()");
+
+            lock (this.consensusLock)
             {
-                result.Error = ex.ConsensusError;
+                try
+                {
+                    this.ValidateBlock(new ContextInformation(item, this.Validator.ConsensusParams));
+                }
+                catch (ConsensusErrorException ex)
+                {
+                    item.Error = ex.ConsensusError;
+                }
+
+                if (item.Error != null)
+                {
+                    this.logger.LogError("Block rejected: {0}", item.Error.Message);
+
+                    // Pull again.
+                    this.Puller.SetLocation(this.Tip);
+
+                    if (item.Error == ConsensusErrors.BadWitnessNonceSize)
+                    {
+                        this.logger.LogInformation("You probably need witness information, activating witness requirement for peers.");
+                        this.connectionManager.AddDiscoveredNodesRequirement(NodeServices.NODE_WITNESS);
+                        this.Puller.RequestOptions(TransactionOptions.Witness);
+                        return;
+                    }
+
+                    // Set the chain back to ConsensusLoop.Tip.
+                    this.Chain.SetTip(this.Tip);
+                    this.logger.LogTrace("Chain reverted back to block '{0}' ", this.Tip);
+
+                    // Since ChainHeadersBehavior check PoW, MarkBlockInvalid can't be spammed.
+                    this.logger.LogError("Marking block as invalid.");
+                    this.chainState.MarkBlockInvalid(item.Block.GetHash());
+                }
+                else
+                {
+                    this.logger.LogTrace("Block accepted '{0}' ", this.Tip);
```

Comment:
I would prefer `Block '{0}' accepted.`, otherwise `Block accepted: '{0}'` with colon and no space at the end